File: src/main/java/io/github/jsnimda/common/config/IConfigOptionPrimitive.java
package io.github.jsnimda.common.config;

import java.util.Objects;

import com.google.gson.JsonElement;
import com.google.gson.JsonPrimitive;

public interface IConfigOptionPrimitive<T> extends IConfigOption {

  T getValue();

  T getDefaultValue();

  void setValue(T value);

  @Override
  default boolean isModified() {
    return !Objects.equals(getValue(), getDefaultValue());
  }

  @Override
  default void resetToDefault() {
    setValue(getDefaultValue());
  }

  @Override
  default JsonElement toJsonElement() {
    T v = getValue();
    if (v instanceof Boolean) {
      return new JsonPrimitive((Boolean)v);
    }
    if (v instanceof Number) {
      return new JsonPrimitive((Number)v);
    }
    if (v instanceof String) {
      return new JsonPrimitive((String)v);
    }
    throw new UnsupportedOperationException("Not implemented yet");
  }

  @Override
  default void fromJsonElement(JsonElement element) {
    if (element.isJsonPrimitive()) {
      JsonPrimitive p = element.getAsJsonPrimitive();
      T v = getDefaultValue();
      if (v instanceof Boolean) {
        if (p.isBoolean()) {
          @SuppressWarnings("unchecked") // TODO find other workarounds than @SuppressWarnings
          T o = (T)(Boolean)p.getAsBoolean();
          setValue(o);
          return;
        }
      } else if (v instanceof Number) {
        if (p.isNumber()) {
          if (v instanceof Integer) {
            @SuppressWarnings("unchecked")
            T o = (T)(Integer)p.getAsInt();
            setValue(o);
          } else if (v instanceof Double) {
            @SuppressWarnings("unchecked")
            T o = (T)(Double)p.getAsDouble();
            setValue(o);
          } else {
            throw new UnsupportedOperationException("Not implemented yet");
          }
          return;
        }
      } else if (v instanceof String) {
        if (p.isString()) {
          @SuppressWarnings("unchecked")
          T o = (T)p.getAsString();
          setValue(o);
          return;
        }
      } else {
        throw new UnsupportedOperationException("Not implemented yet");
      }
    }
    // TODO log fail
  }

}

File: src/main/java/io/github/jsnimda/common/config/IConfigOptionPrimitive.kt
package io.github.jsnimda.common.config

import com.google.gson.JsonElement
import com.google.gson.JsonPrimitive
import io.github.jsnimda.common.Log

interface IConfigOptionPrimitive<T : Any> : IConfigOption {
  var value: T
  val defaultValue: T

  @JvmDefault override val isModified get() = value != defaultValue

  @JvmDefault override fun resetToDefault() {
    value = defaultValue
  }

  @JvmDefault override fun toJsonElement(): JsonElement = when (val v = value) {
    is Boolean -> JsonPrimitive(v)
    is Number -> JsonPrimitive(v)
    is String -> JsonPrimitive(v)
    else -> throw UnsupportedOperationException("Not implemented yet")
  }

  @JvmDefault override fun fromJsonElement(element: JsonElement) {
    resetToDefault()
    try {
      val p = element.asJsonPrimitive
      @Suppress("UNCHECKED_CAST")
      value = when (val v = defaultValue) {
        is Boolean -> p.asBoolean as T
        is Number -> when (v) {
          is Int -> p.asInt as T
          is Double -> p.asDouble as T
          else -> throw UnsupportedOperationException("Not implemented yet")
        }
        is String -> p.asString as T
        else -> throw UnsupportedOperationException("Not implemented yet")
      }
    } catch (e: Exception) {
      Log.warn("[invprofiles.common] Failed to set config value for '$key' from the JSON element '$element'")
    }
  }
}